README.md
[![PyPI version](https://badge.fury.io/py/python-mt-st.svg)](https://pypi.org/project/python-mt-st/)

A magnetic tape drive controller ✨🍰✨
=======================================================

![Screenshot](https://user-images.githubusercontent.com/3183314/44708568-09def800-aada-11e8-9a2c-f576c9d8f00f.png)

A Python module for controlling magnetic tape drives under Linux. The module can perform tape drive operation, just like mt variants (mt-st, mt-gnu, etc.) under Unix-like OSs. For example, a user can rewind, forward/backward space, erase the cartridge, and so on. The current implementation is just a wrapper of mt-gnu. In the future, we will move to communicate with
Linux st driver directly via ioctl instead.


## Usage

[TODO]

## Prerequisites

1. Linux/Python 3
2. The current version is wrapping mt-gnu. Make sure mt-gnu is in your execution path.
3. Root privilege is necessary for operating a magnetic tape drive.
4. You will need either a real tape drive or use mhvtl as a simulate tape drive.

## Install

1. ``$ cd [project_home]``
2. ``$ pip install .``

## Tape Drive Configuration

1. You can find out what tape devices you 
have by this [guide](https://www.bacula.org/7.2.x-manuals/en/problems/Testing_Your_Tape_Drive_Wit.html#SECTION00423000000000000000).  
2. Ensuring that the tape modes are properly set by following commands: (Assume your tape drive is in /dev/nst0)
* mt -f /dev/nst0 stoptions buffer-writes async-writes read-ahead
* mt -f /dev/nst0 defblksize 0

## Testing 

1. You need either a real tape drive or [mhvtl](https://github.com/markh794/mhvtl) for simulation.
2. After tape drive is ready, update tape drive information in ```test/tape_conf.py```
3. run test cases by ```$ make test```

## Known Issues
1. mt-st is necessary is you want to use online/offine operations.

changelog.txt
0.2.0-snapshot ()
- Add online/offline functions

0.1.4 (2018-08-28)
- Initial release


tests/mhvtl.py
# -*- coding: utf-8 -*-
import subprocess
import time

from tests import tape_conf as conf


def reset():
    if conf.HIT_MHVTL:
        _reset()
    else:
        print("Not hit mhvtl. let all drive slots empty and all storage slots full by yourself")


def _reset():
    time.sleep(10)
    subprocess.run(["/etc/init.d/mhvtl", "shutdown"], timeout=10, check=True)
    time.sleep(10)
    subprocess.run(["/etc/init.d/mhvtl", "start"], timeout=10, check=True)
    time.sleep(3)
    # /dev/tape/by-id/* may not generated by udev after restart mhvtl on my machine. Trigger it manually.
    subprocess.run(["udevadm", "trigger"], timeout=10, check=True)
    time.sleep(1)



TIMEOUT = 60 * 3


def unload(device, slotnum: int, drivenum: int, timout=TIMEOUT):
    subprocess.run(['mtx', "-f", device, "unload", str(slotnum), str(drivenum)], timeout=TIMEOUT,
                   check=False)
    print('wait for unload...')
    time.sleep(1)  # mhvtl: unload is asynchronous

tests/tape_conf.py
# -*- coding: utf-8 -*-
import os

# Assumes hitting mhvtl simulated STK L80 medium changer
STK_L80_prop = {
    'mediumx': '/dev/tape/by-id/scsi-SSTK_L80_XYZZY_B',
    'device': '/dev/tape/by-id/scsi-350223344ab000900-nst',
    'first_data_slot_addr': 500,
    'first_storage_slot_addr': 1000,
    'first_storage_slot_cartridge': 'G03001TA',
}

HIT_MHVTL = True

if os.environ.get('HIT_MHVTL') == '0':
    HIT_MHVTL = False

print('HITMHVTL: ' + str(HIT_MHVTL))

"""
Enther first three members
"""
prop = {
    'mediumx': '',
    'device': '',
    'first_storage_slot_cartridge': '',
    'first_data_slot_addr': 0,
    'first_storage_slot_addr': 0,
}

if HIT_MHVTL:
    prop = STK_L80_prop
else:
    print('Make sure config prop by yourself')


tests/test_tape_drive.py
# -*- coding: utf-8 -*-
import subprocess
import time
from unittest import TestCase

from pymtst.tape_drive import TapeDrive
from tests import mhvtl
from tests import tape_conf

test_dir = "/tmp/test_drive"
prop = tape_conf.prop
TIMEOUT = 60 * 3


def prepare_tape():
    if tape_conf.HIT_MHVTL:
        mhvtl.reset()

    subprocess.run(['mtx', "-f", prop['mediumx'], "unload", "1", "0"], timeout=TIMEOUT, check=False)
    print('wait for unload...')
    time.sleep(1)  # mhvtl: unload is asynchronous

    subprocess.run(['mtx', "-f", prop['mediumx'], "load", "1", "0"], timeout=TIMEOUT, check=True)

    if tape_conf.HIT_MHVTL:
        # mhvtl bug mhvtl/issues/4
        subprocess.run(['mt-gnu', "-f", prop['device'], "rewind"], timeout=TIMEOUT, check=True)

    subprocess.run(['mt-gnu', "-f", prop['device'], "weof"], timeout=TIMEOUT, check=True)


class TestTapeDriveOffline(TestCase):
    d = None

    @classmethod
    def setUpClass(self):
        prepare_tape()
        self.d = TapeDrive(prop['device'])

    def test_offline(self):
        self.assertTrue(self.d.is_online())
        self.d.offline()
        self.assertFalse(self.d.is_online())


class TestTapeDrive(TestCase):
    d = None

    @classmethod
    def setUpClass(self):
        prepare_tape()
        self.d = TapeDrive(prop['device'])

    def test_status(self):
        self.assertIsNotNone(self.d.status())

    def test_rewind(self):
        subprocess.run(['mt-gnu', "-f", prop['device'], "eom"], timeout=TIMEOUT, check=True)
        self.d.rewind()
        self.assertEqual(0, self.d.current_file_number())

    def test_eom(self):
        subprocess.run(['mt-gnu', "-f", prop['device'], "rewind"], timeout=TIMEOUT, check=True)
        self.d.eom()
        self.assertNotEqual(0, self.d.current_file_number())

    def test_asf(self):
        subprocess.run(['mt-gnu', "-f", prop['device'], "rewind"], timeout=TIMEOUT, check=True)
        self.d.asf(1)
        self.assertEqual(1, self.d.current_file_number())

        subprocess.run(['mt-gnu', "-f", prop['device'], "eom"], timeout=TIMEOUT, check=True)
        self.d.asf(1)
        self.assertEqual(1, self.d.current_file_number())

        with self.assertRaises(subprocess.CalledProcessError) as cm:
            self.d.asf(999)

        self.assertIn("Input/output error", str(cm.exception.stderr))

    def test_erase(self):
        subprocess.run(['mt-gnu', "-f", prop['device'], "weof"], timeout=TIMEOUT, check=True)
        self.d.erase()
        subprocess.run(['mt-gnu', "-f", prop['device'], "eom"], timeout=TIMEOUT, check=True)
        self.assertEqual(1, self.d.current_file_number())

    def test_current_file_number(self):
        subprocess.run(['mt-gnu', "-f", prop['device'], "rewind"], timeout=TIMEOUT, check=True)
        self.assertEqual(0, self.d.current_file_number())


pymtst/tape_drive.py
# -*- coding: utf-8 -*-
import logging
import subprocess
from functools import wraps
from multiprocessing import Lock

logger = logging.getLogger(__name__)

# This table is based on "HPE StoreEver LTO-8 Ultrium Tape Drives Technical Reference Manual Volume 2 Software
# Integration Guide."
timeouts = {
    'status': 1 * 60,
    'rewind': 11 * 60,
    'eom': 49 * 60,
    'asf': 49 * 60,
    'weof': 28 * 60,
    'erase': 14.8 * 60 * 60,
}

# Undefined in doc. Use rewind's instead
timeouts['offline'] = timeouts['rewind']
timeouts['online'] = timeouts['rewind']

def synchronized(tlockname):
    """A decorator to place an instance based lock around a method """

    def _synched(func):
        @wraps(func)
        def _synchronizer(self, *args, **kwargs):
            tlock = self.__getattribute__(tlockname)
            tlock.acquire()
            try:
                return func(self, *args, **kwargs)
            finally:
                tlock.release()

        return _synchronizer

    return _synched


class TapeDrive:
    """
    Class to represent a magnetic tape drive.

    User can do control operations and actually read/write/erase the cartridge. This implementation is based on
    mt-gnu and python tarfile.

    Note that user should only operate the drive this class represented to via this class simultaneously to prevent
    potential race conditions.

    """

    MT = 'mt-gnu'
    blocking_factor = 1024

    def __init__(self, device, self_check=False):
        """
         :param device: file name of the tape drive to operate on.

        """

        if not str(device).__contains__('n'):
            logger.warning("fool-proofing: it's not a good idea using rewind device here.")

        self.device = device

        if self_check:
            try:
                # fix mhvtl bug
                subprocess.run([self.MT, "-f", self.device, "rewind"], timeout=5, check=True, stderr=subprocess.PIPE,
                               stdout=subprocess.PIPE, universal_newlines=True).stdout

                subprocess.run([self.MT, "-f", self.device, "status"], timeout=5, check=True, stderr=subprocess.PIPE,
                               stdout=subprocess.PIPE, universal_newlines=True).stdout
            # Expect to raise another exception when the drive is busy or unexpected error occurred.
            except subprocess.TimeoutExpired:
                logger.error('Cannot check status... is the cartridge already loaded?')
                raise

        self._lock = Lock()

    def _execute_mt(self, args):
        """
        Execute mt-gnu command with the drive and args

         :param args: an array of mt's "operation, [count]"
         :raise subprocess exception with stderr

        """
        # TODO: wall clock
        logger.debug("pre-execute_mt:: args: '%s', timeout: '%d', file number: '%d'", str(args), timeouts.get(args[0]),
                     self.current_file_number())
        try:
            subprocess.run([self.MT, "-f", self.device] + args, timeout=timeouts.get(args[0]), check=True,
                           stderr=subprocess.PIPE)
        finally:
            logger.debug("post-execute_mt:: file number: '%d'", self.current_file_number())

    @synchronized('_lock')
    def status(self):
        return self._status()

    def _status(self):
        return subprocess.run([self.MT, "-f", self.device, "status"], timeout=timeouts.get("status"), check=True,
                              stderr=subprocess.PIPE,
                              stdout=subprocess.PIPE, universal_newlines=True).stdout

    @synchronized('_lock')
    def rewind(self):
        self._rewind()

    def _rewind(self):
        self._execute_mt(["rewind"])

    @synchronized('_lock')
    def eom(self):
        self._execute_mt(["eom"])

    @synchronized('_lock')
    def asf(self, count):
        self._execute_mt(["asf", str(count)])

    @synchronized('_lock')
    def erase(self, quickly=True):
        self._rewind()

        if quickly:
            self._execute_mt(["weof"])
        else:
            self._execute_mt(["erase"])

    def current_file_number(self):
        return self._parse_file_number(self._status())

    # online/offline operations need mt-st.
    # TODO: move all operations to mt-st.
    @synchronized('_lock')
    def online(self):
        args = ["load"]
        subprocess.run(['mt-st', "-f", self.device] + args, timeout=timeouts.get(args[0]), check=True,
                       stderr=subprocess.PIPE)

    @synchronized('_lock')
    def offline(self):
        args = ["offline"]
        subprocess.run(['mt-st', "-f", self.device] + args, timeout=timeouts.get(args[0]), check=True,
                       stderr=subprocess.PIPE)

    @synchronized('_lock')
    def is_online(self) -> bool:

        status = subprocess.run(['mt-st', "-f", self.device, "status"], timeout=timeouts.get("status"), check=True,
                                stderr=subprocess.PIPE,
                                stdout=subprocess.PIPE, universal_newlines=True).stdout

        return self._word_exist_in_status('ONLINE', status)

    @staticmethod
    def _word_exist_in_status(word, status) -> bool:
        for line in status.splitlines():
            if word in line:
                return True
        return False

    @staticmethod
    def _parse_file_number(status):
        for line in status.splitlines():
            if 'file number' in line:
                logger.debug('file number line: %s', line)
                return int(line.split('=')[1].strip())
        raise Exception('Cannot find file number in status.')
